Sources/swift-parser-test/DiagnosticsFormatter.swift

```diff
@@ -0,0 +1,133 @@
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+
+import Foundation
+import SwiftParser
+import SwiftSyntax
+import SwiftDiagnostics
+
+
+// To put `Diagnostic`s into a set they must conform to hashable.
+// Two `Diagnostic` instances are the same, if they contain the same diagnosticID at the same node.
+extension Diagnostic: Hashable {
+  public func hash(into hasher: inout Hasher) {
+    hasher.combine(self.node)
+  }
+  
+  public static func ==(lhs: Diagnostic, rhs: Diagnostic) -> Bool {
+    return lhs.node == rhs.node && lhs.diagnosticID == rhs.diagnosticID
+  }
+}
+
+struct DiagnosticsFormatter {
+  
+  /**
+   A wrapper struct for a source line and its diagnostics
+   */
+  private struct SourceLine {
+    var diagnostics = Set<Diagnostic>()
+    var sourceString: String
+  }
+  
+  /**
+   Number of lines which should be printed before and after the diagnostic message
+   */
+  static let contextSize = 2
+  
+  /**
+   Print given diagnostics for a given syntax tree on the command line
+   */
+  static func printDiagnostics(tree: SourceFileSyntax, diags: [Diagnostic]) {
+    
+    let slc = SourceLocationConverter(file: "", tree: tree)
+    
+    // First, we need to put each line and its diagnostics together
+    var sourceLines = [SourceLine]()
+    
+    for sourceLine in tree.description.split(separator: "\n", omittingEmptySubsequences: false) {
+      
+      let diagsForLine = diags.filter { diag in
+        let line = diag.location(converter: slc).line
+        return line == (sourceLines.count + 1)
+      }
+      sourceLines.append(SourceLine(diagnostics: .init(diagsForLine), sourceString: String(sourceLine)))
+    }
+    
+    // Only lines with diagnostic messages should be printed, but including some context
+    let rangesToPrint = sourceLines.enumerated().compactMap { (index, sourceLine) -> Range<Int>? in
+      if sourceLine.diagnostics.isEmpty == false {
+        return Range<Int>(uncheckedBounds: (lower: index - Self.contextSize, upper: index + Self.contextSize))
+      }
+      return nil
+    }
+    
+    /// Keep track if a line missing char should be printed
+    var printLineMissingChar = false
+    /// Keep track if the first block of source lines was already printed
+    var isFirstLinePrinted = false
+    
+    for (lineNumber, line) in sourceLines.enumerated() {
+      guard rangesToPrint.contains(where: { range in
+        range.contains(lineNumber)
+      }) else {
+        printLineMissingChar = true
+        continue
+      }
+      
+      // line numbers should be right aligned
+      let numberOfDigits = Int(floor(log10(Float(abs(sourceLines.count))))) + 1
```

Comment:
I moved the declaration up, but had to remove the `log10()`, because that's only available in Foundation. So I changed the calculation to a simpler, uncleaner variant.